value/structreflect.go

```diff
@@ -0,0 +1,238 @@
+	// field might be nested within 'inline' structs
+	for _, elem := range f.fieldPath {
+		structVal = structVal.FieldByIndex(elem)
+	}
+	return structVal
+}
+
+func getStructCacheEntry(t reflect.Type) structCacheEntry {
+	if hints, ok := reflectStructCache.Get(t); ok {
+		return hints
+	}
+
+	hints := map[string]*fieldCacheEntry{}
+	buildStructCacheEntry(t, hints, nil)
+
+	reflectStructCache.Update(t, hints)
+	return hints
+}
+
+func buildStructCacheEntry(t reflect.Type, infos map[string]*fieldCacheEntry, fieldPath [][]int) {
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		jsonName, isInline, isOmitempty := lookupJsonTags(field)
+		if isInline {
+			buildStructCacheEntry(field.Type, infos, append(fieldPath, field.Index))
+			continue
+		}
+		info := &fieldCacheEntry{isOmitEmpty: isOmitempty, fieldPath: append(fieldPath, field.Index)}
+		infos[jsonName] = info
+
+	}
+}
+
+type structReflect struct {
+	Value reflect.Value
+}
+
+func (r structReflect) Length() int {
+	i := 0
+	eachStructField(r.Value, func(s string, value reflect.Value) bool {
+		i++
+		return true
+	})
+	return i
+}
+
+func (r structReflect) Get(key string) (Value, bool) {
+	if val, ok := r.findJsonNameField(key); ok {
+		return mustWrapValueReflect(val), true
+	}
+	return nil, false
+}
+
+func (r structReflect) Has(key string) bool {
+	_, ok := r.findJsonNameField(key)
+	return ok
+}
+
+func (r structReflect) Set(key string, val Value) {
+	fieldVal, ok := r.findJsonNameField(key)
+	if !ok {
+		panic(fmt.Sprintf("key %s may not be set on struct %T: field does not exist", key, r.Value.Interface()))
+	}
+	if !fieldVal.CanSet() {
+		// See https://blog.golang.org/laws-of-reflection for details on why a struct may not be settable
+		panic(fmt.Sprintf("key %s may not be set on struct: %T: struct is not settable", key, r.Value.Interface()))
+	}
+	fieldVal.Set(reflect.ValueOf(val.Unstructured()))
+}
+
+func (r structReflect) Delete(key string) {
+	fieldVal, ok := r.findJsonNameField(key)
+	if !ok {
+		panic(fmt.Sprintf("key %s may not be deleted on struct %T: field does not exist", key, r.Value.Interface()))
+	}
+	if !fieldVal.CanSet() {
+		// See https://blog.golang.org/laws-of-reflection for details on why a struct may not be settable
+		panic(fmt.Sprintf("key %s may not be deleted on struct: %T: struct is not settable", key, r.Value.Interface()))
+	}
+	fieldVal.Set(reflect.Zero(fieldVal.Type()))
+}
+
+func (r structReflect) Iterate(fn func(string, Value) bool) bool {
+	return eachStructField(r.Value, func(s string, value reflect.Value) bool {
+		v := mustWrapValueReflect(value)
+		defer v.Recycle()
+		return fn(s, v)
+	})
+}
+
+func eachStructField(structVal reflect.Value, fn func(string, reflect.Value) bool) bool {
+	for jsonName, fieldCacheEntry := range getStructCacheEntry(structVal.Type()) {
+		fieldVal := fieldCacheEntry.getFieldFromStruct(structVal)
+		if fieldCacheEntry.isOmitEmpty && (safeIsNil(fieldVal) || isZero(fieldVal)) {
```

Comment:
I think callers expect to visit every field in the *schema* whether or not it is set in the given object? @apelisse @jennybuckley

Reply:
No I don't think so. In the unstructured world we unmarshal into a map[string]interface{} and only know about the specified fields.

Reply:
That was my understanding as well.

In a subsequent PR, I'll be merging a test that I have in k8s/k8s to go through all native kubernetes types, fuzz them, wrap them with this reflect value interface, convert them to unstructured and wrap them with the unstructured value interface and make sure they match. That's passing for me locally.